Patch Tuesday for November 2024

BillyONeal opened the pull request:
* Move Azure Region WestUS3 -> WestUS
* Update Visual Studio to 17.11.6.
* Update the Windows SDK to the current version. This required working around https://developercommunity.visualstudio.com/t/NAN-is-no-longer-compile-time-constant-i/10688907 in a few ports
* Update PowerShell to 7.4.6.
* Update Azure-CLI to 2.66.0.
* Remove the WDK as this appears to be unused.
* Synchronize the Linux image minting JSON with as-deployed.
* Upgrade from `PublishBuildArtifacts@1` to `PublishPipelineArtifact@1`
* Add `arm64-windows-static-md` as requested by Azure Core.
* Fixed `[chartdir]`'s supports expression to match its portfile.cmake.
* Fixed `[freeopcua]` needing /bigobj and combined patches of patches.
* Enabled ports skipped for not having 'stable download URLs' now that we have asset caching.

BillyONeal committed:
Update VS to 17.11.6, and remove some older components we shouldn't be using.

BillyONeal committed:
Update pwsh to v7.4.6

BillyONeal committed:
Update Azure-CLI to 2.66.0

BillyONeal committed:
Remove the WDK.

BillyONeal committed:
Add Linux apt update (as already deployed).

BillyONeal committed:
Upgrade Android docker container to Noble.

BillyONeal committed:
Move to WestUS

BillyONeal committed:
More move to westus.

BillyONeal committed:
Upgrade to PublishPipelineArtifact.

BillyONeal committed:
Update image.

BillyONeal commented:
I forgot to make the Azure DevOps changes. Will deploy that and reopen.

BillyONeal committed:
More move to westus.

BillyONeal committed:
Promote arm64-windows-static-md as requested by Azure Core.

BillyONeal commented:
/azp run

Neumann-A commented on scripts/azure-pipelines/windows/deploy-visual-studio.ps1:
```diff
@@ -22,9 +22,6 @@ $Workloads = @(
   'Microsoft.VisualStudio.Component.VC.ATL.ARM64',
   'Microsoft.VisualStudio.Component.VC.MFC.ARM64',
   'Microsoft.VisualStudio.Component.Windows11SDK.22621',
```
```suggestion
  'Microsoft.VisualStudio.Component.Windows11SDK.26100',
```

BillyONeal replied:
Not sure if that one is present in 17.11.6's installer. Will certainly do this when we move to 17.12.2 or later.

BillyONeal commented on scripts/azure-pipelines/windows/deploy-visual-studio.ps1:
```diff
@@ -22,9 +22,6 @@ $Workloads = @(
   'Microsoft.VisualStudio.Component.VC.ATL.ARM64',
   'Microsoft.VisualStudio.Component.VC.MFC.ARM64',
   'Microsoft.VisualStudio.Component.Windows11SDK.22621',
-  'Microsoft.VisualStudio.Component.Windows10SDK.20348',
```
Looks like this needs to get put back because https://gitlab.kitware.com/cmake/cmake/-/issues/25232

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into patch-tuesday-2024-11

BillyONeal committed:
[freeopcua] Fix arm64-windows build, combine boost patches, minimize patches.

BillyONeal committed:
Duplicate all arm64-windows ci.baseline.txt entries as arm64-windows-static-md.

BillyONeal committed:
Upgrade Windows 11 SDK as suggested by @Neumann-A , and put Windows 10 SDK back.

BillyONeal committed:
Update Windows image again.

BillyONeal committed:
Try to improve reliability by switching to curl.exe rather than Invoke-WebRequest.

BillyONeal committed:
Revert the Android update for now.

BillyONeal commented:
Something in that Android image is broken, so I reverted this part for now and will investigate next week:

* Update the Android base Docker container to Ubuntu 24.04 "Noble Numbat", unblocks https://github.com/microsoft/vcpkg/pull/40996

BillyONeal commented on scripts/azure-pipelines/windows/utility-prefix.ps1:
```diff
@@ -64,7 +64,7 @@ Function DownloadAndInstall {
       New-Item -ItemType Directory -Path $tempPath -Force | Out-Null
       $LocalPath = Join-Path $tempPath $LocalName
-      Invoke-WebRequest -Uri $Url -OutFile $LocalPath
+      curl.exe -L -o $LocalPath $Url
```
Invoke-WebRequest kept dying with funny errors like "a remote connection was forcibly closed" seemingly randomly. I don't know if this will fix it, but it passed after 3 tries failed.

BillyONeal committed:
[freeopcua] Fix version DB

BillyONeal committed:
[chartdir] Fix supports expression to match the actual prebuilt DLLs available.

BillyONeal committed:
[igraph] Workaround non-constant NAN.

BillyONeal committed:
[libgit2] Remove workaround for old cmake that breaks arm64-windows-static.

Not submitted upstream because it isn't clear they care more about old cmake than arm64, but we do.

BillyONeal committed:
[librttopo] Workaround Windows SDK non-constant NAN.

BillyONeal committed:
Bring back .NET bits, and apply ci.baseline.txt edits.

BillyONeal commented:
>REGRESSION: chartdir:x64-android failed with BUILD_FAILED. If expected, add chartdir:x64-android=fail to /vcpkg/scripts/azure-pipelines/../ci.baseline.txt.

This also appears in the contemporary build for this and I'm not sure why it only appears here.

```
Building chartdir:x64-android@7.0.0#7...
-- Note: chartdir only supports dynamic library linkage. Building dynamic library.
CMake Error at ports/chartdir/portfile.cmake:93 (file):
  file INSTALL cannot find "//LICENSE.TXT": No such file or directory.
Call Stack (most recent call first):
  scripts/ports.cmake:192 (include)


error: building chartdir:x64-android failed with: BUILD_FAILED
```

Given that this port downloads a prebuilt DLL, the supports expression should match exactly what portfile.cmake does, which I have changed it to do.

> REGRESSION: cppmicroservices:arm64-windows-static-md failed with BUILD_FAILED. If expected, add cppmicroservices:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

Similar to existing arm64-windows skip.

> REGRESSION: dcmtk:arm64-windows-static-md failed with BUILD_FAILED. If expected, add dcmtk:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

I typo'd an =fail

> REGRESSION: easyhook:x64-windows failed with BUILD_FAILED. If expected, add easyhook:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: easyhook:x86-windows failed with BUILD_FAILED. If expected, add easyhook:x86-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

Needs .NET framework bits :(

> REGRESSION: glibmm:arm64-windows-static-md failed with BUILD_FAILED. If expected, add glibmm:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: libsoundio:arm64-windows-static-md failed with BUILD_FAILED. If expected, add libsoundio:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: libtomcrypt:arm64-windows-static-md failed with BUILD_FAILED. If expected, add libtomcrypt:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: netcdf-cxx4:arm64-windows-static-md failed with BUILD_FAILED. If expected, add netcdf-cxx4:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: openmama:arm64-windows-static-md failed with BUILD_FAILED. If expected, add openmama:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: zyre:arm64-windows-static-md failed with BUILD_FAILED. If expected, add zyre:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

Similar to existing skips.

> REGRESSION: lcm:arm64-windows-static-md failed with BUILD_FAILED. If expected, add lcm:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

Similar to existing x64-windows-static-md skip, looks like incorrect CRT settings.

> REGRESSION: igraph:arm64-uwp failed with BUILD_FAILED. If expected, add igraph:arm64-uwp=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: igraph:arm64-windows failed with BUILD_FAILED. If expected, add igraph:arm64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: igraph:arm64-windows-static-md failed with BUILD_FAILED. If expected, add igraph:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: igraph:x64-uwp failed with BUILD_FAILED. If expected, add igraph:x64-uwp=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: igraph:x64-windows failed with BUILD_FAILED. If expected, add igraph:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: igraph:x64-windows-static failed with BUILD_FAILED. If expected, add igraph:x64-windows-static=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: igraph:x64-windows-static-md failed with BUILD_FAILED. If expected, add igraph:x64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: igraph:x86-windows failed with BUILD_FAILED. If expected, add igraph:x86-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: librttopo:arm64-uwp failed with BUILD_FAILED. If expected, add librttopo:arm64-uwp=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: librttopo:arm64-windows failed with BUILD_FAILED. If expected, add librttopo:arm64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: librttopo:arm64-windows-static-md failed with BUILD_FAILED. If expected, add librttopo:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: librttopo:x64-uwp failed with BUILD_FAILED. If expected, add librttopo:x64-uwp=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: librttopo:x64-windows failed with BUILD_FAILED. If expected, add librttopo:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: librttopo:x64-windows-static failed with BUILD_FAILED. If expected, add librttopo:x64-windows-static=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: librttopo:x64-windows-static-md failed with BUILD_FAILED. If expected, add librttopo:x64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: librttopo:x86-windows failed with BUILD_FAILED. If expected, add librttopo:x86-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

https://developercommunity.visualstudio.com/t/NAN-is-no-longer-compile-time-constant-i/10688907

> REGRESSION: libgit2:arm64-windows-static-md failed with BUILD_FAILED. If expected, add libgit2:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

Old CMake workaround hardcodes /MACHINE:x64, removed here.

> REGRESSION: openni2:x64-windows failed with BUILD_FAILED. If expected, add openni2:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: openni2:x64-windows-static-md failed with BUILD_FAILED. If expected, add openni2:x64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: openni2:x86-windows failed with BUILD_FAILED. If expected, add openni2:x86-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

```
Installing 3/3 openni2:x64-windows@2.2.0.33#15...
Building openni2:x64-windows@2.2.0.33#15...
CMake Error at ports/openni2/portfile.cmake:3 (message):
  Could not find <cor.h>.  Ensure the NETFXSDK is installed.
```

also needs .NET

> REGRESSION: openimageio:arm64-windows-static-md failed with BUILD_FAILED. If expected, add openimageio:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: openmvs:arm64-windows-static-md failed with BUILD_FAILED. If expected, add openmvs:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

New triplet, does not repro locally, and the cause of failure isn't obvious. The failure looks like it's actually in opencv4, possible missing find_dependency /cc @cenit

```
CMake Error at D:/installed/arm64-windows-static-md/share/opencv4/OpenCVModules.cmake:355 (set_target_properties):
  The link interface of target "opencv_wechat_qrcode" contains:

    Iconv::Iconv

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  D:/installed/arm64-windows-static-md/share/opencv4/OpenCVConfig.cmake:126 (include)
  D:/installed/arm64-windows-static-md/share/opencv/vcpkg-cmake-wrapper.cmake:3 (_find_package)
  D:/a/_work/1/s/scripts/buildsystems/vcpkg.cmake:813 (include)
  src/cmake/checked_find_package.cmake:132 (find_package)
  src/cmake/externalpackages.cmake:207 (checked_find_package)
  CMakeLists.txt:176 (include)
```

> REGRESSION: openmvg:arm64-windows-static-md failed with BUILD_FAILED. If expected, add openmvg:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

```
C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1441~1.341\bin\Hostx64\arm64\cl.exe   /TP  -ID:\installed\arm64-windows-static-md\include\eigen3 -ID:\b\openmvg\src\1047522df9-a2b2f63b2c.clean\src -external:ID:\installed\arm64-windows-static-md\include -external:W0 /nologo /DWIN32 /D_WINDOWS /utf-8 /GR /EHsc /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  /nologo /DWIN32 /D_WINDOWS /utf-8 /GR /EHsc /MP -D_USE_MATH_DEFINES /showIncludes /FoopenMVG\matching_image_collection\CMakeFiles\openMVG_matching_image_collection.dir\Matcher_Regions.cpp.obj /FdopenMVG\matching_image_collection\CMakeFiles\openMVG_matching_image_collection.dir\openMVG_matching_image_collection.pdb /FS -c D:\b\openmvg\src\1047522df9-a2b2f63b2c.clean\src\openMVG\matching_image_collection\Matcher_Regions.cpp
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.41.34120\include\nmmintrin.h(17): fatal error C1189: #error:  This header is specific to X86, X64, ARM64, and ARM64EC targets
```

This looks kind of like https://github.com/opencv/opencv/issues/24235 ? New triplet though, so just skipped.

> REGRESSION: sentencepiece:arm64-windows-static-md failed with BUILD_FAILED. If expected, add sentencepiece:arm64-windows-static-md=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> REGRESSION: sentencepiece:x64-windows-static failed with BUILD_FAILED. If expected, add sentencepiece:x64-windows-static=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.

Fixed by https://github.com/microsoft/vcpkg/pull/42205

> REGRESSION: cmake-user:arm64-windows cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: cmake-user:arm64-windows-static-md cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: cmake-user:x64-windows cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: cmake-user:x64-windows-static cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: cmake-user:x64-windows-static-md cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: cmake-user:x86-windows cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-gdal:x64-windows cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-gdal:x64-windows-static cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-gdal:x64-windows-static-md cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-gdal:x86-windows cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-opencv:x64-windows cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-opencv:x64-windows-static cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-opencv:x64-windows-static-md cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-openimageio:x64-windows cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-openimageio:x64-windows-static cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-openimageio:x64-windows-static-md cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-paraview:x64-windows cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-paraview:x64-windows-static cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).
> REGRESSION: vcpkg-ci-paraview:x64-windows-static-md cascaded, but it is required to pass. (D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt).

I'm assuming that these are caused by one of the other failures for now.

BillyONeal committed:
Update Windows images.

BillyONeal commented:
>##[error]Artifact cannot be uploaded because max quantity has been exceeded or the payment instrument is invalid. https://aka.ms/artbilling for details.

I asked on an internal mailing list if this is expected.

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into patch-tuesday-2024-11

BillyONeal committed:
[gdal] Remove /machine flag.

BillyONeal committed:
Fix another ci.baseline.txt typo.

BillyONeal committed:
Also OpenMVS

BillyONeal committed:
Version DB

BillyONeal commented:
/azp run

BillyONeal commented:
> > ##[error]Artifact cannot be uploaded because max quantity has been exceeded or the payment instrument is invalid. https://aka.ms/artbilling for details.
> 
> I asked on an internal mailing list if this is expected.

I fixed this by 'setting up billing' but given that pipeline artifacts are documented to be unbillable it seems like a bug in either the product or the docs, I reported.

BillyONeal commented:
Closed and reopened to kick CLABot

BillyONeal commented on ports/freeopcua/boost-compatibility.patch:
```diff
@@ -31,8 +57,30 @@ index 795a6b2..532ac3c 100644
+   , Callback(callback)
+   , io(service.GetIOService())
+-  , Timer(io, boost::posix_time::milliseconds(data.RevisedPublishingInterval))
++  , Timer(io, boost::posix_time::milliseconds((int)data.RevisedPublishingInterval))
```
This part used to be in ports/freeopcua/improve_compatibility_with_recent_boost.patch

BillyONeal commented on ports/freeopcua/cmakelists_fixes.patch:
```diff
@@ -68,11 +68,12 @@ index de03564..df3fcf6 100644
 -    #set(CMAKE_SHARED_LINKER_FLAGS "--no-undefined" )
 +    add_definitions(-D_SCL_SECURE_NO_WARNINGS)
 +    add_definitions(-D_CRT_SECURE_NO_WARNINGS)
++    add_compile_options(/bigobj)
```
The main reason I touched this port was to add this `/bigobj` because that's what the comment in `ci.baseline.txt` said it needed, and that appears to be correct :)

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into patch-tuesday-2024-11

Neumann-A commented:
@BillyONeal: Any tips to make easyhook work? I simply cannot get it to pass the post build checks. What exactly needs to be installed by VS?

BillyONeal commented:
> @BillyONeal: Any tips to make easyhook work? I simply cannot get it to pass the post build checks. What exactly needs to be installed by VS?

It's currently working in CI. When it failed before it complained about .NET targeting stuff not being installed.

Neumann-A commented:
> It's currently working in CI. When it failed before it complained about .NET targeting stuff not being installed.

Did it really complain about that or did it fail because the EasyHook.dll does not export any symbols?

BillyONeal commented:
> Did it really complain about that or did it fail because the EasyHook.dll does not export any symbols?

It failed because easyhook explicitly asks for .NET things in its build system.

BillyONeal commented:
```console
MSBuild version 17.11.9+a69bbaaf5 for .NET Framework
Build started 11/15/2024 3:36:06 PM.

     1>Project "D:\b\easyhook\x64-windows-rel\EasyHook.sln" on node 1 (EasyHookDll target(s)).
     1>ValidateSolutionConfiguration:
         Building solution configuration "netfx4-Release|x64".
       ValidateProjects:
         The project "EasyHook.Tests" is not selected for building in solution configuration "netfx4-Release|x64".
     1>Project "D:\b\easyhook\x64-windows-rel\EasyHook.sln" (1) is building "D:\b\easyhook\x64-windows-rel\EasyHookDll\EasyHookDll.vcxproj" (2) on node 1 (default targets).
     2>Project "D:\b\easyhook\x64-windows-rel\EasyHookDll\EasyHookDll.vcxproj" (2) is building "D:\b\easyhook\x64-windows-rel\EasyHookSvc\EasyHookSvc.csproj" (3:2) on node 1 (default targets).
     3>C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(1259,5): error MSB3644: The reference assemblies for .NETFramework,Version=v4.7.2 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks [D:\b\easyhook\x64-windows-rel\EasyHookSvc\EasyHookSvc.csproj]
     3>_CleanRecordFileWrites:
         Creating directory "obj\netfx4-Release\".
     3>Done Building Project "D:\b\easyhook\x64-windows-rel\EasyHookSvc\EasyHookSvc.csproj" (default targets) -- FAILED.
     2>Project "D:\b\easyhook\x64-windows-rel\EasyHookDll\EasyHookDll.vcxproj" (2) is building "D:\b\easyhook\x64-windows-rel\EasyHook\EasyHook.csproj" (4:2) on node 2 (default targets).
     4>C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(1259,5): error MSB3644: The reference assemblies for .NETFramework,Version=v4.7.2 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks [D:\b\easyhook\x64-windows-rel\EasyHook\EasyHook.csproj]
     4>_CleanRecordFileWrites:
         Creating directory "obj\netfx4-Release\".
     4>Done Building Project "D:\b\easyhook\x64-windows-rel\EasyHook\EasyHook.csproj" (default targets) -- FAILED.
     2>Done Building Project "D:\b\easyhook\x64-windows-rel\EasyHookDll\EasyHookDll.vcxproj" (default targets) -- FAILED.
     1>Done Building Project "D:\b\easyhook\x64-windows-rel\EasyHook.sln" (EasyHookDll target(s)) -- FAILED.

Build FAILED.

       "D:\b\easyhook\x64-windows-rel\EasyHook.sln" (EasyHookDll target) (1) ->
       "D:\b\easyhook\x64-windows-rel\EasyHookDll\EasyHookDll.vcxproj" (default target) (2) ->
       "D:\b\easyhook\x64-windows-rel\EasyHookSvc\EasyHookSvc.csproj" (default target) (3:2) ->
       (GetReferenceAssemblyPaths target) -> 
         C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(1259,5): error MSB3644: The reference assemblies for .NETFramework,Version=v4.7.2 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks [D:\b\easyhook\x64-windows-rel\EasyHookSvc\EasyHookSvc.csproj]


       "D:\b\easyhook\x64-windows-rel\EasyHook.sln" (EasyHookDll target) (1) ->
       "D:\b\easyhook\x64-windows-rel\EasyHookDll\EasyHookDll.vcxproj" (default target) (2) ->
       "D:\b\easyhook\x64-windows-rel\EasyHook\EasyHook.csproj" (default target) (4:2) ->
         C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(1259,5): error MSB3644: The reference assemblies for .NETFramework,Version=v4.7.2 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks [D:\b\easyhook\x64-windows-rel\EasyHook\EasyHook.csproj]

    0 Warning(s)
    2 Error(s)

Time Elapsed 00:00:00.71
```
